File: server/internal/models/canteen.go
package models

import (
	"time"

	"gorm.io/gorm"
)

const (
	// CanteenOperatingActive 表示食堂正常营业，可参与新评价、排行和推荐。
	CanteenOperatingActive = "active"
	// CanteenOperatingOffline 表示食堂暂时下架，历史数据保留但不参与当前发现体系。
	CanteenOperatingOffline = "offline"
)

// 食堂位置系统标签：区域（一食堂/二食堂）与楼层（一楼/二楼）。
// 店铺名只保留纯店铺名，位置信息由这两个结构化字段表达。
const (
	CanteenLocationArea1  = "一食堂"
	CanteenLocationArea2  = "二食堂"
	CanteenLocationFloor1 = "一楼"
	CanteenLocationFloor2 = "二楼"
)

// CanteenLocationAreas 区域标签可选值。新增食堂区域时在此扩展。
var CanteenLocationAreas = []string{CanteenLocationArea1, CanteenLocationArea2}

// CanteenLocationFloors 楼层标签可选值。新增楼层时在此扩展。
var CanteenLocationFloors = []string{CanteenLocationFloor1, CanteenLocationFloor2}

// IsValidCanteenLocationArea 校验区域标签是否为合法可选值（空值表示未填写）。
func IsValidCanteenLocationArea(v string) bool {
	if v == "" {
		return true
	}
	for _, area := range CanteenLocationAreas {
		if area == v {
			return true
		}
	}
	return false
}

// IsValidCanteenLocationFloor 校验楼层标签是否为合法可选值（空值表示未填写）。
func IsValidCanteenLocationFloor(v string) bool {
	if v == "" {
		return true
	}
	for _, floor := range CanteenLocationFloors {
		if floor == v {
			return true
		}
	}
	return false
}

// Canteen 食堂/店铺
type Canteen struct {
	ID              uint       `gorm:"primaryKey" json:"id"`
	Name            string     `gorm:"size:100;not null;index" json:"name"`
	NormalizedName  string     `gorm:"size:100;not null;default:''" json:"-"`
	LocationArea    string     `gorm:"size:20;not null;default:''" json:"location_area"`
	LocationFloor   string     `gorm:"size:20;not null;default:''" json:"location_floor"`
	Image           string     `gorm:"size:500;not null" json:"image"` // 封面图
	Verified        bool       `gorm:"default:false" json:"verified"`  // 仅管理员审核通过后公开
	OperatingStatus string     `gorm:"size:20;not null;default:'active';index" json:"operating_status"`
	OfflinedAt      *time.Time `json:"offlined_at,omitempty"`
	OfflinedBy      *uint      `json:"-"`
	OfflineReason   string     `gorm:"size:500" json:"-"`
	IsOffline       bool       `gorm:"-" json:"is_offline"`
	CreatedBy       uint       `gorm:"index" json:"created_by"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`

	// 提交奖励经验状态：审核通过发放、下架/删除收回、恢复营业返还。
	ExpRewardedAt *time.Time `json:"-"`
	ExpRevokedAt  *time.Time `json:"-"`

	RatingCount int     `gorm:"-" json:"rating_count"`
	AverageStar float64 `gorm:"-" json:"average_star"`

	// 管理端待审核列表返回的提交人昵称（非数据库字段）
	CreatorName string `gorm:"-" json:"creator_name,omitempty"`
}

// NormalizeOperatingStatus 为历史数据库补齐状态，并计算只读 JSON 字段。
func (c *Canteen) NormalizeOperatingStatus() {
	if c.OperatingStatus == "" {
		c.OperatingStatus = CanteenOperatingActive
	}
	c.IsOffline = c.OperatingStatus == CanteenOperatingOffline
}

// EnsureCanteenOperatingStatusSchema 是幂等的历史数据补齐迁移。
func EnsureCanteenOperatingStatusSchema(db *gorm.DB) error {
	return db.Model(&Canteen{}).
		Where("operating_status IS NULL OR operating_status = ''").
		Update("operating_status", CanteenOperatingActive).Error
}

// CanteenRating 食堂评价
type CanteenRating struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	CanteenID      uint      `gorm:"index;not null" json:"canteen_id"`
	UserID         uint      `gorm:"index;not null" json:"user_id"`
	Star           int       `gorm:"not null" json:"star"`                      // 1-5星
	Comment        string    `gorm:"size:500" json:"comment"`                   // 评价文字
	Images         string    `gorm:"type:text" json:"images"`                   // 图片JSON数组
	Tags           string    `gorm:"type:text" json:"tags"`                     // 体验标签JSON数组
	HelpfulCount   int       `gorm:"not null;default:0" json:"helpful_count"`   // 有用票数
	UnhelpfulCount int       `gorm:"not null;default:0" json:"unhelpful_count"` // 无用票数
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	Status         string    `gorm:"size:20;not null;default:'active';index" json:"status"`

	// V2 摘要字段。旧客户端继续读取 Star；新版使用 EffectiveScore 与五维摘要。
	EffectiveScore      float64 `gorm:"not null;default:0" json:"effective_score"`
	TasteScore          float64 `gorm:"not null;default:0" json:"taste_score"`
	ValueScore          float64 `gorm:"not null;default:0" json:"value_score"`
	QueueScore          float64 `gorm:"not null;default:0" json:"queue_score"`
	HygieneScore        float64 `gorm:"not null;default:0" json:"hygiene_score"`
	ServiceScore        float64 `gorm:"not null;default:0" json:"service_score"`
	ReviewEventCount    int     `gorm:"not null;default:0" json:"review_event_count"`
	LatestReviewEventID *uint   `gorm:"index" json:"latest_review_event_id,omitempty"`
	ScoreVersion        int     `gorm:"not null;default:1" json:"score_version"`

	// 关联数据（非数据库字段）
	User                 *User    `gorm:"foreignKey:UserID" json:"-"`
	UserName             string   `gorm:"-" json:"user_name"`
	UserStudentID        string   `gorm:"-" json:"user_student_id"`
	UserAvatar           string   `gorm:"-" json:"user_avatar"`
	MyVote               *string  `gorm:"-" json:"my_vote"`
	RecommendedDishNames []string `gorm:"-" json:"recommended_dishes,omitempty"`
	CreditScore          int      `gorm:"-" json:"credit_score,omitempty"`
	CreditWeight         float64  `gorm:"-" json:"credit_weight,omitempty"`
	HistoryCount         int      `gorm:"-" json:"history_count,omitempty"`
}

// CanteenRatingVote 食堂评价有用/无用投票
type CanteenRatingVote struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	RatingID  uint      `gorm:"not null;uniqueIndex:idx_canteen_rating_vote_user" json:"rating_id"`
	UserID    uint      `gorm:"not null;uniqueIndex:idx_canteen_rating_vote_user" json:"user_id"`
	VoteType  string    `gorm:"type:varchar(10);not null" json:"vote_type"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

File: client/lib/models/canteen.dart
import 'dart:convert';

class Canteen {
  final int id;
  final String name;
  final String locationArea;
  final String locationFloor;
  final String image;
  final bool verified;
  final String operatingStatus;
  final int createdBy;
  final int ratingCount;
  final int reviewerCount;
  final int visitReviewCount;
  final double averageStar;
  final Map<String, double> dimensionScores;
  final int dishCount;
  final int dishWithPhotoCount;
  final int dishPhotoCount;
  final double rankingScore;

  Canteen({
    required this.id,
    required this.name,
    this.locationArea = '',
    this.locationFloor = '',
    required this.image,
    required this.verified,
    this.operatingStatus = 'active',
    required this.createdBy,
    required this.ratingCount,
    this.reviewerCount = 0,
    this.visitReviewCount = 0,
    required this.averageStar,
    this.dimensionScores = const {},
    this.dishCount = 0,
    this.dishWithPhotoCount = 0,
    this.dishPhotoCount = 0,
    this.rankingScore = 0,
  });

  factory Canteen.fromJson(Map<String, dynamic> json) {
    return Canteen(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      locationArea: json['location_area']?.toString() ?? '',
      locationFloor: json['location_floor']?.toString() ?? '',
      image: json['image'] ?? '',
      verified: json['verified'] ?? false,
      operatingStatus: json['is_offline'] == true
          ? 'offline'
          : (json['operating_status']?.toString() ?? 'active'),
      createdBy: json['created_by'] ?? 0,
      ratingCount: json['rating_count'] ?? 0,
      reviewerCount: json['reviewer_count'] ?? json['rating_count'] ?? 0,
      visitReviewCount: json['visit_review_count'] ?? json['rating_count'] ?? 0,
      averageStar: (json['average_star'] ?? 0).toDouble(),
      dimensionScores: (json['dimension_scores'] is Map)
          ? (json['dimension_scores'] as Map).map(
              (key, value) =>
                  MapEntry(key.toString(), (value as num?)?.toDouble() ?? 0),
            )
          : const {},
      dishCount: json['dish_count'] ?? 0,
      dishWithPhotoCount: json['dish_with_photo_count'] ?? 0,
      dishPhotoCount: json['dish_photo_count'] ?? 0,
      rankingScore: (json['ranking_score'] ?? 0).toDouble(),
    );
  }

  bool get isOffline => operatingStatus == 'offline';

  /// 位置标签文案，如"一食堂·二楼"；未填写时为空。
  String get locationLabel {
    if (locationArea.isEmpty && locationFloor.isEmpty) return '';
    if (locationArea.isEmpty) return locationFloor;
    if (locationFloor.isEmpty) return locationArea;
    return '$locationArea·$locationFloor';
  }
}

class CanteenRating {
  final int id;
  final int canteenId;
  final int userId;
  final int star;
  final String comment;
  final List<String> images;
  final String userName;
  final String userStudentId;
  final String userAvatar;
  final String createdAt;
  final int helpfulCount;
  final int unhelpfulCount;
  final String? myVote;
  final int tasteScore;
  final int valueScore;
  final int queueScore;
  final int hygieneScore;
  final int serviceScore;
  final int creditScore;
  final double creditWeight;
  final int historyCount;

  CanteenRating({
    required this.id,
    required this.canteenId,
    required this.userId,
    required this.star,
    required this.comment,
    required this.images,
    required this.userName,
    required this.userStudentId,
    required this.userAvatar,
    required this.createdAt,
    this.helpfulCount = 0,
    this.unhelpfulCount = 0,
    this.myVote,
    this.tasteScore = 0,
    this.valueScore = 0,
    this.queueScore = 0,
    this.hygieneScore = 0,
    this.serviceScore = 0,
    this.creditScore = 0,
    this.creditWeight = 0,
    this.historyCount = 0,
  });

  factory CanteenRating.fromJson(Map<String, dynamic> json) {
    List<String> parsedImages = [];
    if (json['images'] != null && json['images'].toString().isNotEmpty) {
      try {
        final decoded = jsonDecode(json['images']);
        if (decoded is List) {
          parsedImages = List<String>.from(decoded);
        }
      } catch (e) {
        // json decode failed
      }
    }
    return CanteenRating(
      id: json['id'] ?? 0,
      canteenId: json['canteen_id'] ?? 0,
      userId: json['user_id'] ?? 0,
      star: json['star'] ?? 0,
      comment: json['comment'] ?? '',
      images: parsedImages,
      userName: json['user_name'] ?? '匿名',
      userStudentId: json['user_student_id'] ?? '',
      userAvatar: json['user_avatar'] ?? '',
      createdAt: json['created_at'] ?? '',
      helpfulCount: json['helpful_count'] ?? 0,
      unhelpfulCount: json['unhelpful_count'] ?? 0,
      myVote: json['my_vote']?.toString(),
      tasteScore: (json['taste_score'] ?? 0).toInt(),
      valueScore: (json['value_score'] ?? 0).toInt(),
      queueScore: (json['queue_score'] ?? 0).toInt(),
      hygieneScore: (json['hygiene_score'] ?? 0).toInt(),
      serviceScore: (json['service_score'] ?? 0).toInt(),
      creditScore: (json['credit_score'] ?? 0).toInt(),
      creditWeight: (json['credit_weight'] ?? 0).toDouble(),
      historyCount: (json['history_count'] ?? 0).toInt(),
    );
  }
}
